runtime/envelope/envelope.go

```diff
@@ -254,100 +152,121 @@ func (e *Envelope) runWeavelet(ctx context.Context) error {
 	cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", runtime.ToEnvelopeKey, strconv.FormatUint(uint64(toEnvelopeFd), 10)))
 	cmd.Env = append(cmd.Env, e.config.Env...)
 
-	// Different sources are read from by different go-routines.
-	var stdoutErr, stderrErr, weaveletConnErr error
-	var wait sync.WaitGroup
-	wait.Add(2) // 2 for stdout and stderr
-	go func() {
-		defer wait.Done()
-		stdoutErr = e.copyLines("stdout", outpipe)
-	}()
+	// Start the command.
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// Capture stdout and stderr from the weavelet.
+	// TODO(spetrovic): These need to be terminated and their errors taken into
+	// account. Fix along with fixing the Stop() behavior.
 	go func() {
-		defer wait.Done()
-		stderrErr = e.copyLines("stderr", errpipe)
+		if err := e.copyLines("stdout", outpipe); err != nil {
+			fmt.Fprintf(os.Stderr, "Error copying stdout: %v\n", err)
+		}
 	}()
-	wait.Add(1)
 	go func() {
-		defer wait.Done()
-		weaveletConnErr = conn.Run()
+		if err := e.copyLines("stderr", errpipe); err != nil {
+			fmt.Fprintf(os.Stderr, "Error copying stdout: %v\n", err)
+		}
 	}()
 
-	// Start the command.
-	if err := cmd.Start(); err != nil {
+	// Create the connection, now that the weavelet has (hopefully) started.
+	conn, err := conn.NewEnvelopeConn(toEnvelope, toWeavelet, e.handler, e.weavelet)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Unable to start envelope conn: %v\n", err)
 		return err
 	}
-	e.mu.Lock()
-	e.process = cmd.Process
-	e.mu.Unlock()
 
-	// Set the connection only after the weavelet information was sent to the
-	// subprocess. Otherwise, it is possible to send over the pipe information
-	// (e.g., health checks) before the weavelet information was sent.
-	e.setConn(conn)
+	e.cmd = cmd
+	e.conn = conn
+	return nil
+}
+
+// toggleProfiling compares the value of e.profiling to the given expected
+// value, and if they are the same, toggles the value of e.profiling and
+// returns true; otherwise, it leaves the value of e.profiling unchanged
+// and returns false.
+func (e *Envelope) toggleProfiling(expected bool) bool {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.profiling != expected {
+		return false
+	}
+	e.profiling = !e.profiling
+	return true
+}
 
-	// Wait for the command to terminate.
-	//
-	// NOTE(mwhittaker): Stop also calls Wait. If Stop is called, then the
-	// redundant wait leads to a "waitid: no child processes" error which we
-	// ignore.
-	wait.Wait()
-	runErr := cmd.Wait()
-	for _, err := range []error{runErr, stdoutErr, stderrErr, weaveletConnErr} {
+// Serve blocks accepting incoming messages from the weavelet.
+func (e *Envelope) Serve(ctx context.Context) error {
```

Comment:
I was a bit surprised that the `ctx` argument to `Serve` was not used to control the lifetime of this method. I guess the expectation is that either `Envelope.Stop` should be called to stop this, or the `context` passed to `NewEnvelope` should be cancelled. Would it be appropriate to remove this `ctx` argument since it's ignored?

Reply:
Yeah, that's the state I'd like to end up in really soon. Will try to make the stopping/cancellation context based in a followup cl.